backend/error_source_test.go

```diff
@@ -16,3 +18,44 @@ func TestErrorSource(t *testing.T) {
 	require.True(t, backend.ErrorSourcePlugin.IsValid())
 	require.Equal(t, "plugin", backend.ErrorSourcePlugin.String())
 }
+
+func TestResponseWithOptions(t *testing.T) {
+	for _, tc := range []struct {
+		name            string
+		err             error
+		expErrorMessage string
+		expErrorSource  backend.ErrorSource
+	}{
+		{
+			name:            "plugin error",
+			err:             backend.PluginError(errors.New("unknown")),
+			expErrorMessage: "unknown",
+			expErrorSource:  backend.ErrorSourcePlugin,
+		},
+		{
+			name:            "downstream error",
+			err:             backend.DownstreamError(errors.New("bad gateway")),
+			expErrorMessage: "bad gateway",
+			expErrorSource:  backend.ErrorSourceDownstream,
+		},
+		{
+			name:            "wrapped downstream error",
+			err:             fmt.Errorf("wrapped: %w", backend.DownstreamError(errors.New("inside error"))),
+			expErrorMessage: "wrapped: inside error",
+			expErrorSource:  backend.ErrorSourceDownstream,
+		},
+		{
+			name:            "wrapped plugin error",
+			err:             fmt.Errorf("wrapped: %w", backend.PluginError(errors.New("inside error"))),
+			expErrorMessage: "wrapped: inside error",
+			expErrorSource:  backend.ErrorSourcePlugin,
+		},
```

Comment:
Maybe another case here for completeness?
```go
		{
			name:            "non error source error",
			err:             errors.New("inside error"),
			expErrorMessage: "inside error",
			expErrorSource:  "",
		},
```

Reply:
Good idea! Added.